fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/BottomToolbarContainerView.kt

```diff
@@ -33,6 +38,8 @@ class BottomToolbarContainerView(
     navigationItems: List<ActionItem> = NavigationItems.defaultItems,
     androidToolbarView: View? = null,
     menuButton: MenuButton,
+    browserStore: BrowserStore,
+    browsingModeManager: BrowsingModeManager,
```

Comment:
After @MatthewTighe's recent PR, you may be able to just directly observe the browsing mode from `AppState` (if that is ready to be consumed on this dev branch).

Reply:
In fact, the `BrowsingModeManager` was removed completely so this change will be required once you rebase the dev branch onto main.

Reply:
Here is the patch for context https://github.com/mozilla-mobile/firefox-android/pull/4646

Reply:
thanks! I think I will deal with it later on, I assume there will be more conflicts when I am about to merge this into main

Reply:
Have yall been rebasing at all since making the dev branch?

Reply:
I haven't yet

Reply:
I think you might be doing disservice to yourself by not rebasing your dev branch with main and continuing forward with your current PR in its current state. Addressing merge conflicts before landing back to main does mean your original reviewed PRs will now contain changes that haven't been reviewed and will probably raise issues down the line compare to getting it right now. 

As an aside, I am wondering with this dev branch that we have agreed to use as rapid prototyping to reach some findings on how we should proceed with implementing this project - has there been any discovery or decision reached on how we should proceed? Eg, was this gonna to be done in XML or a mixed of XML/Compose where possible? 

If y'all believe there is a good path forward and there are any share outs that you can provide the team to help validate that this is a good way to move forward. I would recommend we move towards landing your changes back onto main if the investigative work has been productive and has support from the team.

If we are still prototyping, then feel free to continue with the development branch.

Reply:
@gabrielluong 
That's a very good point, [here](https://docs.google.com/document/d/1Lo76-tYhkgB-8LZh5nefaOxouIPFTo_x2x8w2_VPbhU/edit) it is. I was actually going to ask you to take a look. I don't intend to wait for the whole team approval, but everybody is welcomed to make suggestions/raise concerns. 

I was actually approaching the dev branch as something that eventually will become a patch to merge into main.

Reply:
Before I do my final pass, what is the verdict here? When will the dev branch be going away?

I don't have anything more to add about the perils of not rebasing a dev branch with main periodically; GL's comment summarized my thoughts/concerns.

Reply:
I get the concerns. My intention was to open up a PR for a review, to merge the dev branch into the main. It's not the most optimal way to do it, but it would have a benefit of commits being previously reviewed so it won't have surprises during the final review. I was going to rebase before this final feature version.

I did rebase it now to speed up the process, but ironically Matt's change has been reverted.

Reply:
I think it's worthwhile to highlight some risks with how we approach with the project and areas where we want to make sure we are using the best judgement:

- Given this would've been a multi-month development branch, it would be highly encouraged that we continue to rebase the dev branch to avoid bitrot and leaving rebase and merge conflicts at the end. This was probably the most notable red flag here. This isn't a point against using the development branch, but more about how we are  mitigating risk in our day-to-day development.
- This particular patch has some upstream implications because it is essentially refactoring the tab counter UI that is in-use today. I think it would be better to consider landing the part of the refactoring on main and then rebasing the change back into the development branch. This will require some judgement on your part in your development cycle because ideally we keep the isolated feature inside of the development branch. Think of the development branch as where you can prototype quickly, keep a feature in complete secret until it's ready to merge and ship or if you're working on a redesign/change of an existing feature that cannot be feature flag easily because it builds on top of something existing. The point is the development branch is to keep all the feature development separate and isolated from main. Given how the current PR can be beneficial upstream, I would highly encourage landing this upstream to make it easier for everyone and rebasing the development branch with your changes.
- When we initially discussed the development branch, we were still trying to prototype quickly between using Compose and XML or both, and ensure we only landed what was needed in main at the end. How ya'll been reviewing and landing incremental changes into the development branch to build the new feature has been what we wanted to see. Looking at your doc, I think the next steps would be to communicate the implementation details and get buy-in from a domain expert on the browser toolbar in terms of how we are implementing this. I don't really know if I am the domain expert here, but I also recognize that such a person might not exist on the team anymore. The likely candidate I can think of is Jon or Roger.
- There is a couple of things that have changed with this project since we started and I think it's time to start thinking about how we can mitigate some of these risks in the coming month. One reasoning for this is that your primary reviewer on this project will be going on leave soon. Any new reviewer will not have the context of the previous patches that will be landing back into the main, which will cause its own set of issues. Another concern is that the monorepo migration will likely happen soon. Timing wise I think Noah provides a good call out for re-considering when we should merge back into main. It already looks like the development branch has reached a big milestone if it is clear to y'all how the implementation needs to be approached and you can get buy-in from someone who is familiar with the existing browser toolbar.

My proposed next steps:

1. Let's land the tab counter UI refactoring into main and then we can have a different PR that uses this new element in the dev branch.
2. Find a domain expert (this could be 1-3 people, not everyone needs to be an expert, but ideally one person is familiar with the toolbar) around the browser toolbar and start communicating your implementation plan. Aim to find a reviewer who can take over next month in this process.
3. Communicate a plan on how you want to proceed with the dev branch and the pathway back to main. We can absolutely continue with the dev branch in the short term and nothing needs to change. By Feb, we will need to have an idea of how we will mitigate the concerns that I have listed above. My recommendation is to aim to reland this back in main earlier once we have established that the implementation plan is sound. An alternative idea would be to plan for an appropriate milestone checkpoint (maybe when the feature no longer needs the incomplete feature flag) as when you decide to reland back to main. You will need to work out the timing of this because this involves people's schedule for leave, how long you think this need to be in development for, and monorepo migration timeline.

Reply:
We had a follow up with @gabrielluong, takeaways are
1. [landed](https://github.com/mozilla-mobile/firefox-android/pull/5365) refactoring separately 
2. gl will be taking a closer look into the project, effectively taking a domain expert role. We lost Jonathan, who had a lot of context about the toolbar, to another team, the navigation bar change doesn't affect the toolbar itself a lot, but rather is adding new compose elements while reusing the existing code,- given all of that I am very happy to see him helping out with it.
3. I am planning to open a PR with the most feature next week, which should give us enough time before Harrison going on parental leave and monorepo migration. In other words, I expect to merge a few smaller patches for dev branch, and then the feature is ready for main.